pkg/proxy/ipvs/proxier.go

```diff
@@ -2007,6 +2022,19 @@ func (proxier *Proxier) syncEndpoint(svcPortName proxy.ServicePortName, onlyNode
 		}
 
 		if curEndpoints.Has(ep) {
+			// if we want to update weights, loop through all current destinations and
+			// reset their weight.
+			if proxier.updateWeights {
+				for _, dest := range curDests {
```

Comment:
Add `proxier.updateWeights = false` here. The mutex is held so it's safe

Reply:
(btw, you may make a comment that the mutex is held)

Reply:
we can't add this here, as this is inside a loop. adding this outside the loop instead.

Reply:
Outside the "for _, ep := range newEndpoints.List() {" loop you mean? Seens reasonable.

Reply:
on second thought, is it completely safe to have `proxier.updateWeights = false` inside `syncProxyRules()` instead of `OnServicesSynced()`/`OnEndpointSynced()`?
i'm asking because in `OnServicesSynced()` we do something like:
```golang
proxier.mu.Lock()
...
proxier.updateWeights = true
proxier.mu.Unlock()

proxier.syncProxyRules()
```
so even though `proxier.syncProxyRules()` would capture the mutex first thing, theoretically another goroutine which called `syncProxyRules()` could execute `proxier.updateWeights = false` at the exact moment between `OnServiceSynced` let go of the lock and `proxier.syncProxyRules()` acquired the lock.

Reply:
But it should go into the "syncProxyRules()" function somewhere, else you got a race since the lock is released and reclaimed. The proxier.updateWeights = false should be done when holding the lock in syncProxyRules().

Reply:
You are right. To set proxier.updateWeights = false in "syncProxyRules()" makes it safer, but setting proxier.updateWeights = false outside "syncProxyRules()" and release the lock is not safe.

A safe way may be
1. Initiate the proxy struct with someBool = true
2. In "syncProxyRules()" (lock held), do the one-time stuff and set someBool = false

However this would require 2 bools

Reply:
then i'm wondering whether it makes sense to set `proxier.updateWeights = true` _only_ in `OnServicesSynced()` and not on `OnEndpointSliceSynced()`, since we only need to sync the endpoints once.

Reply:
`OnServicesSynced()` makes more sense than `OnEndpointSliceSynced()` because that's guaranteed to run once on startup: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/server.go#L746, unlike `OnEndpointSliceSynced()`, which may not be run if the cluster is using only `Endpoint` and not `EndpointSlice`

Reply:
I was about to ask that :smile: If the function only needs to be called once (not one time for endpoints and one time for endpointslices), then initiate "updateWeights" to true when the proxier struct is created and set it to false immediately after "if proxier.updateWeights"

Reply:
yes, that works as well. let's go ahead with that.

Reply:
Is this possible?
```go
func (proxier *Proxier) syncProxyRules() {
	proxier.mu.Lock()
	defer proxier.mu.Unlock()
        defer proxier.updateWeights = false
```
If so, a generic bool like `firstTimeSync` may be appropriate and may be used by future one-shot functions.

Reply:
yes, this is better 👍

Reply:
I made a test but it's not working yet https://go.dev/play/p/y-HwPYW3eNa

Reply:
Silly mistake, now it works

Reply:
yes, i noticed `first` needs to be initialized to true instead of false

Reply:
Seems `go` has support for "Once"; https://golangcode.com/run-code-once-with-sync/

You may see if you can use it.

Reply:
I am unsure if it's better since we already have the proxier.mu mutex.

Reply:
I think it's better to avoid `sync.Once` because if in the future we want to do more things only on startup, the underlying function will get large. Having a flag gives us flexibility about where we want to run the one-stop logic during sync.